turbinia/pubsub.py

```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
     self.__dict__ = obj
 
 
+class TurbiniaCelery(object):
+  """Celery app object.
+
+  Attributes:
+    app (Celery): The Celery app itself.
+    _fexec: Lets us initialize the Celery task after configuring the app, since
+        the Celery 'app.task' decorator is unknown before initialization.
+    fexec (function, args, kwargs): Celery task, which can run functions
+        without needing to pre-register them (similar to PSQ). Takes in a
+        function name, as well as any arguments (named or otherwise). All
+        workers must have this function defined.
+  """
+
+  def __init__(self):
+    """Celery configurations."""
+    self.app = None
+    self.fexec = None
+
+  def _fexec(self):
+    """Closure used to pass functions to workers."""
+    @self.app.task(name='fexec')
+    def fexec(f, *args, **kwargs):
+      """Lets us pass in an arbitrary function without Celery annotations"""
+      return f(*args, **kwargs)
+    return fexec
+
+  def setup(self):
+    """Set up Celery"""
+    config.LoadConfig()
+    self.app = celery.Celery(
+        'turbinia',
+        broker=config.CELERY_BROKER,
+        backend=config.CELERY_BACKEND
+    )
+    self.app.conf.update(
+        task_default_queue=config.INSTANCE_ID,
+        event_serializer='pickle',
+        result_serializer='pickle',
+        task_serializer='pickle',
+        accept_content=['pickle'],
+        # TODO(ericzinnikas): Without task_acks_late Celery workers will start
+        # on one task and prefetch another (i.e. can result in 1 worker getting
+        # 2 plaso jobs while another worker is free). But enabling this causes
+        # problems with certain Celery brokers (duplicated work).
+        task_acks_late=False,
+        task_track_started=True,
+        worker_concurrency=1,
+        worker_prefetch_multiplier=1,
+    )
+    self.fexec = self._fexec()
```

Comment:
Agreed it is a bit weird, since `fexec` is set to the return value of the `_fexec` closure (which is another function). I'll work on updating the docstring and creating a placeholder method that documents the functionality.

Reply:
thanks